src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -370,34 +356,24 @@ def activeTask(doc, spec):
         # for each expectedRSEs not in localCopyOfCurrentRSEs
         for rse in set(expectedRSEs).difference(localCopyOfCurrentRSEs):
             # check whether the RSE has enough available space available for that
-            if isinstance(rucioClient, Client):
-                generator = rucioClient.get_rse_usage(rse)
-            else:
-                generator = rucioClient.getRSEUsage(rse)
+            # taken from logic of RSEQuotas, see https://bit.ly/3kDJZmO
+            # but here we use explicitly rse parameter
+            records = rucioClient.getAccountUsage(rucioAccount, rse)
             enoughSpace = False
-            for rec in generator:
-                # use nodeUsage provided in spec to get rse space info
-                if nodeUsage and rse in nodeUsage:
-                    rseUsage = nodeUsage[rse]
-                    if rseUsage['bytes_remaining'] > marginSpace:
-                        enoughSpace = True
-                        break
-                else:
-                    # we'll use total - used space to exceed pileupSize with some margin
-                    if rec['total'] - rec['used'] - pileupSize > marginSpace:
-                        enoughSpace = True
-                        break
+            for rec in records:
```

Comment:
How about implementing a method in RSEQuotas, where you provide the rseName and the pileupSize, and it says whether the RSE has enough space or not (of course, considering the marginSpace that is passed when constructing the RSEQuotas object).
If you implement this, please make a unit test as well.

Reply:
If you'll inspect RSEQuotas code it does exactly what I put here, i.e. a call to Rucio Wrapper and store stuff in internal cache. I'm not in favor of wrappers around wrappers around wrappers, see my [talk](https://indico.cern.ch/event/1126680/contributions/5087996/attachments/2530898/4354581/DBSMigration2Go.pdf) slide 12 I made on last C&O week. Here we have rucio client which is wrapper around rucio REST, rucio Wrapper which is wrapper around rucio client, and RSEQuotas which is a wrapper around Rucio wrapper. This overhead brings no new functionality in my view. If you want to make changes to RSEQuotas that's fine by me, but I suggest to make them separately as it delays MSPileup development (as you said new unit test should be done, new review, etc.) Again, the trade-off between common codebase and overhead should be clear. In this case I choose middle layer, i.e. a Rucio wrapper API call which provides exactly what we need.

Reply:
Your slide seem to mention about wrappers at the REST WMCore layer, which is completely different than the Rucio wrapper. Also because you mentioned this in the O&C week, it does not mean that is the absolute truth and way to deal with this.

In addition to that, RSEQuotas is NOT a wrapper around the Rucio wrapper (reason why we don't even import Rucio in the module itself). It's a class to store RSE information and it provides many helper methods. We do not want to implement the same (or even worse, different and potentially wrong) logic for evaluating whether a RSE has enough storage and can accept a data with size XYZ. To me, that's much better than throwing that logic in the middle of this method.

Reply:
Alan, please try to understand what I'm saying. The overhead of wrappers is not specific to REST layer. It is universal.

The RSEQuotas only make one call here https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/MSTransferor/DataStructs/RSEQuotas.py#L115 using rucio Wraper (call dataSvc). Then it stores results into its cache. That's it. If you want to use it In MSPileup we need to:
- add import to RSEQuotas
- initialize RSEQuotas
- adjust activeTask spec to pass this object around
- use the API
- parse results of API

The amount of additional code it will bring to MSPileup is much more than just one call we need using Rucio Wrapper. Because of that the overhead not only in calls, but also in codebase, in inter-module dependencies, etc. But such overhead brings nothing in functionality. If you want to modify RSEQuotas to handle single RSE you need to decide either you'll make a new API (which is again a single call) or modify existing code of fetchStorageUsage API to have extra parameter. In this case the overhead for MSPileup will be memory used by RSEQuotas because it is a cache layer. All of these overheads sum-ups at different level and contribute to service resource consumption, etc.

Again, I'm not arguing when wrapper provides new functionality, but in this case the wrapper brings literally nothing and only brings codebase, dependencies and resource utilization overhead.

I'll not argue anymore, if you still disagree I'll put back the RSEQuotas.

Reply:
Oh, I thought you had already imported RSEQuotas in this module.
Yes, I agree with you, if it brings in more complexity to this module, let's not do it. Thank you for the follow up!